include/fw_info.rst

```diff
@@ -0,0 +1,123 @@
+Every image must provide an ABI getter (:cpp:member:`abi_out`) that other images can use to retrieve its ABIs.
+This ABI getter is a function pointer that retrieves ABI structs (or rather pointers to ABI structs).
+
+In addition, every image can access other ABIs through a second ABI getter (:cpp:member:`abi_in`).
+This ABI getter must be provided when booting the image.
+In other words, an image should expect its :cpp:member:`abi_in` ABI getter to be filled at the time it boots, and will not touch it during booting.
+After booting, an image can call :cpp:member:`abi_in` to retrieve ABIs from the image or images that booted it without knowing where they are located.
+
+Each image can provide multiple ABIs.
+An ABI getter function takes an index, and each index from 0 to *n* must return a different ABI, given that the image provides *n* ABIs with the same ID.
+
+Typically, the actual ABI will be a function pointer (or a list of function pointers), but the data can be anything, though it should be considered read-only.
+The reason for making the ABI getters function pointers instead of pointing to a list of ABIs is to allow dynamic creation of ABI lists without using RAM.
+
+Usage
+*****
+
+To locate and verify firmware info structures, use :cpp:func:`fw_info_find` and :cpp:func:`fw_info_check`, respectively.
+
+To find an ABI with a given version and flags, call :cpp:func:`fw_info_abi_find`.
+This function calls :cpp:member:`abi_in` under the hood, checks the ABI's version against the allowed range, and checks that it has all the flags set.
+
+To populate an image's :cpp:member:`abi_in` (before booting the image), the booting image should call :cpp:func:`fw_info_abi_provide` with its own firmware information structure.
+Note that if the booting (current) firmware image and the booted image's RAM overlap, :cpp:func:`fw_info_abi_provide` will corrupt the current firmware's RAM.
+This is ok if it is done immediately before booting the other image.
+
+Creating ABIs
+*************
+
+To create an ABI, complete the following steps:
+
+1. Declare a new struct type that starts with the :c:type:`fw_info_abi` struct:
+
+   .. code-block:: c
+
+      struct my_abi {
+      	   struct fw_info_abi header;
+   	   struct {
+   		   /* Actual ABI/data goes here. */
+   	   } abi;
+      };
+
+#. Use the :c:macro:`__ext_abi` macro to initialize the ABI struct in an arbitrary location.
+   :c:macro:`__ext_abi` will automatically include the ABI in the list provided via :cpp:func:`fw_info_abi_provide`.
+
+   .. code-block:: c
+
+      __ext_abi(struct my_abi, my_abi) = {
+   	   .header = FW_INFO_ABI_INIT(MY_ABI_ID,
+   				   CONFIG_ABI_FLAGS,
+   				   CONFIG_ABI_VER,
+   				   sizeof(struct my_abi)),
+   	   .abi = {
+   		   /* ABI initialization goes here. */
+   	   }
+      };
+
+#. To include function pointers in your ABI, call the :c:macro:`EXT_ABI_FUNCTION` macro to forward-declare the function and create a typedef for the function pointer:
```

Comment:
I don't understand what we're doing here?

Reply:
A forward declaration is what is usually done in a header file:

    in my_abi_foo(bool arg1, int *arg2);

The `EXT_ABI_FUNCTION()` macro does this, and also creates a typedef for the function. This means that it creates a new type which can hold function pointers to this function. They are declared like this (to create a type called by_abi_foo_t):

    typedef int (*my_abi_foo_t)(bool arg1, int *arg2);

We want this typedef to be able to put ABI function pointers into the ABI struct.
The `EXT_ABI_FUNCTION()` macro is a convenience macro that does both of these so the user doesn't have to type them both, possibly making errors.